File: 2024/day10/cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <queue>
#include <set>
#include <array>

// Directions: up, down, left, right
constexpr std::array<std::array<int, 2>, 4> DIRS = {{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}};

std::vector<std::pair<int, int>> find_trailheads(const std::vector<std::vector<int>>& grid) {
    std::vector<std::pair<int, int>> trailheads;
    const int rows = grid.size();
    const int cols = grid[0].size();

    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (grid[r][c] == 0) {
                trailheads.push_back({r, c});
            }
        }
    }
    return trailheads;
}

int count_reachable_nines(const std::vector<std::vector<int>>& grid, int start_r, int start_c) {
    const int rows = grid.size();
    const int cols = grid[0].size();

    std::set<std::pair<int, int>> visited;
    std::set<std::pair<int, int>> nines;
    std::queue<std::pair<int, int>> q;

    visited.insert({start_r, start_c});
    q.push({start_r, start_c});

    while (!q.empty()) {
        const auto [r, c] = q.front();
        q.pop();

        const int current_height = grid[r][c];

        if (current_height == 9) {
            nines.insert({r, c});
            continue;
        }

        // Try all four directions
        for (const auto& dir : DIRS) {
            const int nr = r + dir[0];
            const int nc = c + dir[1];

            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
                if (visited.find({nr, nc}) == visited.end()) {
                    if (grid[nr][nc] == current_height + 1) {
                        visited.insert({nr, nc});
                        q.push({nr, nc});
                    }
                }
            }
        }
    }

    return nines.size();
}

int dfs(const std::vector<std::vector<int>>& grid, int r, int c) {
    const int rows = grid.size();
    const int cols = grid[0].size();
    const int current_height = grid[r][c];

    if (current_height == 9) {
        return 1;
    }

    int total = 0;
    for (const auto& dir : DIRS) {
        const int nr = r + dir[0];
        const int nc = c + dir[1];

        if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            if (grid[nr][nc] == current_height + 1) {
                total += dfs(grid, nr, nc);
            }
        }
    }

    return total;
}

int part1(const std::vector<std::vector<int>>& grid) {
    const std::vector<std::pair<int, int>> trailheads = find_trailheads(grid);
    int total_score = 0;
    for (const auto [r, c] : trailheads) {
        total_score += count_reachable_nines(grid, r, c);
    }
    return total_score;
}

int part2(const std::vector<std::vector<int>>& grid) {
    const std::vector<std::pair<int, int>> trailheads = find_trailheads(grid);
    int total_rating = 0;
    for (const auto [r, c] : trailheads) {
        total_rating += dfs(grid, r, c);
    }
    return total_rating;
}

int main() {
    std::ifstream input("../input.txt");
    std::string line;
    std::vector<std::vector<int>> grid;

    while (std::getline(input, line)) {
        std::vector<int> row;
        for (char c : line) {
            row.push_back(c - '0');
        }
        grid.push_back(row);
    }

    input.close();

    std::cout << "Part 1: " << part1(grid) << std::endl;
    std::cout << "Part 2: " << part2(grid) << std::endl;

    return 0;
}

File: 2024/day10/perl/solution.pl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10;

# Read input file
my $input_file = "../input.txt";
open(my $fh, '<', $input_file) or die "Cannot open $input_file: $!";
my @lines = <$fh>;
close($fh);
chomp(@lines);

# Parse input into grid
my @grid;
foreach my $line (@lines) {
    my @row = split //, $line;
    push @grid, [map { int($_) } @row];
}

my $rows = scalar @grid;
my $cols = scalar @{$grid[0]};

# Directions: up, down, left, right
my @DIRS = ([-1, 0], [1, 0], [0, -1], [0, 1]);

sub find_trailheads {
    my @trailheads;
    for my $r (0 .. $rows - 1) {
        for my $c (0 .. $cols - 1) {
            if ($grid[$r][$c] == 0) {
                push @trailheads, [$r, $c];
            }
        }
    }
    return @trailheads;
}

sub count_reachable_nines {
    my ($start_r, $start_c) = @_;

    # BFS to find all 9s reachable from a trailhead
    my %visited;
    my @queue = ([$start_r, $start_c]);
    $visited{"$start_r,$start_c"} = 1;
    my %nines;

    while (@queue) {
        my $current = shift @queue;
        my ($r, $c) = @$current;
        my $current_height = $grid[$r][$c];

        if ($current_height == 9) {
            $nines{"$r,$c"} = 1;
            next;
        }

        # Try all four directions
        foreach my $dir (@DIRS) {
            my ($dr, $dc) = @$dir;
            my $nr = $r + $dr;
            my $nc = $c + $dc;

            if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
                my $key = "$nr,$nc";
                if (!exists $visited{$key}) {
                    if ($grid[$nr][$nc] == $current_height + 1) {
                        $visited{$key} = 1;
                        push @queue, [$nr, $nc];
                    }
                }
            }
        }
    }

    return scalar keys %nines;
}

sub count_distinct_trails {
    my ($start_r, $start_c) = @_;

    # DFS to count all distinct trails from a trailhead to any 9
    my $dfs;
    $dfs = sub {
        my ($r, $c) = @_;
        my $current_height = $grid[$r][$c];

        if ($current_height == 9) {
            return 1;
        }

        my $total = 0;
        foreach my $dir (@DIRS) {
            my ($dr, $dc) = @$dir;
            my $nr = $r + $dr;
            my $nc = $c + $dc;

            if ($nr >= 0 && $nr < $rows && $nc >= 0 && $nc < $cols) {
                if ($grid[$nr][$nc] == $current_height + 1) {
                    $total += $dfs->($nr, $nc);
                }
            }
        }

        return $total;
    };

    return $dfs->($start_r, $start_c);
}

sub part1 {
    my @trailheads = find_trailheads();
    my $total_score = 0;
    foreach my $th (@trailheads) {
        my ($r, $c) = @$th;
        $total_score += count_reachable_nines($r, $c);
    }
    return $total_score;
}

sub part2 {
    my @trailheads = find_trailheads();
    my $total_rating = 0;
    foreach my $th (@trailheads) {
        my ($r, $c) = @$th;
        $total_rating += count_distinct_trails($r, $c);
    }
    return $total_rating;
}

# Main execution
say "Part 1: " . part1();
say "Part 2: " . part2();
